keps/sig-api-machinery/1623-standardize-conditions/README.md

```diff
@@ -0,0 +1,144 @@
+milestone **before the [Enhancement Freeze](https://git.k8s.io/sig-release/releases)
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core
+Kubernetes i.e., [kubernetes/kubernetes], we require the following Release
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+- [ ] Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] KEP approvers have approved the KEP status as `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+While many Kuberentes APIs have `.status.conditions`, the schema of `condition` varies a lot between them.
+There is very little commonality at the level of serialization, proto-encoding, and required vs optional.
+Conditions are central enough to the API to make a common golang type with a fixed schema.
+The schema can be a strong recommendation to all API authors.
+
+## Motivation
+
+Allow general consumers to expect a common schema for `.status.conditions` and share golang logic for common Get, Set, Is for `.status.conditions`.
+The pattern is well-established and we have a good sense of the schema we now want.
+
+### Goals
+
+ 1. For all new APIs, have a common type for `.status.conditions`.
+ 2. Provide common utility methods for `HasCondition`, `IsConditionTrue`, `SetCondition`, etc.
+ 3. Provide recommended defaulting functions that set required fields and can be embedded into conversion/default functions.
+
+### Non-Goals
+
+ 1. Update all existing APIs to make use of the new condition type.
+
+## Proposal
+
+Introduce a type into k8s.io/apimachinery/pkg/apis/meta/v1 for `Condition` that looks like 
+```go
+type Condition struct {
+	// Type of condition in CamelCase.
+	// +required
+	Type string `json:"type" protobuf:"bytes,1,opt,name=type"`
+	// Status of the condition, one of True, False, Unknown.
+	// +required
+	Status ConditionStatus `json:"status" protobuf:"bytes,2,opt,name=status"`
```

Comment:
If we're normalizing, should we consider bigger changes - like making this an optional bool?

Reply:
> If we're normalizing, should we consider bigger changes - like making this an optional bool?

I think we've been happy with this as an enumerated string to cover Unknown as well.

Reply:
Who is happy?  All I see are calls to `ParseBool`.  "Unknown" as a value is just weird because we also have to handle a non-specified condition, which is truly unknown.

TL;DR I am not super happy with Conditions.

Reply:
You called this out as unresolved. 

I guess I'm arguing for a larger, deeper, not really compatible change.  

> I think the acknowledgment of unknown has worked well. It gives an option for acknowledging awareness of a condition without the ability to inspect the current value. For instance, an inability to reach an external system results in an Unknown status for the "is-it-installed" condition. You cannot do that with the absence of a condition. I think the concern about ParseBool is well handled by an IsConditionTrue and IsConditionFalse which comments here indicate are commonly used.

If we were rebooting `Condition`, I would argue harder for `*bool` as the totally unambiguous representation.  Given that you're not rebooting, just normalizing, I will waive my objection.  I just wanted it on the record, I guess. :)

Reply:
Given that everyone is already using "Unknown" as a value, I think we have to keep it. :/

Documentation should make clear that "Unknown" is a fact about the *writer of the condition*, and not a claim about *the object*.

Good: `{Ready, Unknown}` to mean: "it might or might not be ready, we don't know."

Good: `{IPAssigned, Unknown}` to mean: "it might or might not have gotten an IP address, we don't know."

Very bad: `{IPAssigned, Unknown}` to mean: "it has an IP address, but we don't know what it is."

Reply:
+1

But also `<no key "IPAssigned">` is equivalent to `{IPAssigned, Unknown}` for most intents.

Reply:
`{IPAssigned, Unknown}` gives a hint to other participants that the `"IPAssigned"` condition may still be reconciled, while `<no key "IPAssigned">` makes it more difficult for other participants to know whether IP assignment is applicable in this case. But this is mostly about the case for `*bool` vs `bool`, rather than the string enum that we have vs what we wish.

Would it be reasonable to fill in `status: ""` as `status: "Unknown"` somewhere in the apiserver side for Conditions (read side or write side)? My gut says "no", but my heart says "please".

Reply:
Yes, I think `status: ""` should be canonicalized as `status: "Unknown"`.

Reply:
> Would it be reasonable to fill in `status: ""` as `status: "Unknown"` somewhere in the apiserver side for Conditions (read side or write side)? My gut says "no", but my heart says "please".

I would require condition writers to say what they mean and specify the status value.

I'd much rather have a condition-adding patch that typos `"statuss":"True"` get rejected with a "status is required" error than get accepted with status defaulted to "Unknown"